uart-npcx: forward poll requests to fifo when running interrupt mode

yperess opened the pull request:
Looking at the data sheet for the npcx section 4.15.5 CR_UART Core
Registers, the UICTRL register that is used in the npcx poll
functions is invalid when running in FIFO mode. Instead, calls to
uart_npcx_poll_in and uart_npcx_poll_out should be redirected to
their respective FIFO read/fill functions (when using interrupts).

Signed-off-by: Yuval Peress <peress@chromium.org>

Commit:
uart-npcx: forward poll requests to fifo when running interrupt mode

Looking at the data sheet for the npcx section 4.15.5 CR_UART Core
Registers, the UICTRL register that is used in the npcx poll
functions is invalid when running in FIFO mode. Instead, calls to
uart_npcx_poll_in and uart_npcx_poll_out should be redirected to
their respective FIFO read/fill functions (when using interrupts).

Changes:
- When calling uart_poll_in: forward the request to uart_fifo_read.
- When calling uart_poll_out: loop until uart_fifo_fill returns
  non-0 (we wrote 1 byte).

Signed-off-by: Yuval Peress <peress@chromium.org>

MulinChao reviewed (approved):
LGTM.

MulinChao commented on drivers/serial/uart_npcx.c:
```diff
@@ -90,11 +90,23 @@ static void uart_npcx_clear_rx_fifo(const struct device *dev)
 /* UART api functions */
 static int uart_npcx_poll_in(const struct device *dev, unsigned char *c)
 {
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
```
One minor comment. If we replace this line with 'if (IS_ENABLED(CONFIG_UART_INTERRUPT_DRIVEN))', does it causes build-error when CONFIG_UART_INTERRUPT_DRIVEN is not set? If not, I think using 'IS_ENABLED()' macro is more readable.

galak replied:
In this case it might be better to just do:
```
#ifdef CONFIG_UART_INTERRUPT_DRIVEN
static int uart_npcx_poll_in(const struct device *dev, unsigned char *c)
{
   .... /* interrupt version */
}
#else
static int uart_npcx_poll_in(const struct device *dev, unsigned char *c)
{
   .... /* poll version */
}
#endif

yperess replied:
@galak I think you're right, I'll just roll these into the above #ifdef.

galak commented on drivers/serial/uart_npcx.c:
```diff
@@ -90,9 +90,41 @@ static void uart_npcx_clear_rx_fifo(const struct device *dev)
 
-/* UART api functions */
+/* Forward declaration to be used in uart_npcx_poll_in(). */
+static int uart_npcx_fifo_read(const struct device *dev, uint8_t *rx_data,
```
Can you re-organize (ie, move `uart_npcx_poll_in` and `uart_npcx_poll_out` so they are after `uart_npcx_fifo_read` and `uart_npcx_fifo_fill`) so we don't need the forward declarations?

yperess replied:
Sure thing

yperess replied:
OK, I took the section below that was also guarded by `CONFIG_UART_INTERRUPT_DRIVEN` and merged it with the one above. I can split this up into 2 commits if it makes it easier to review. Let me know.

galak replied:
Yeah, 2 commits would be better, one that just moves the code and the second that makes the changes.

yperess replied:
Sure, I'll upload ASAP

yperess replied:
Done

fabiobaltieri commented:
Hey do we have any good reason to keep byte mode support at all? Thinking that we may just have the controller set in FIFO mode regardless of CONFIG_UART_INTERRUPT_DRIVEN, so that we don't have to have two different poll implementation to maintain and test for the poll API, as well as a change of behavior if the interrupt mode gets compiled out.

yperess commented:
> Hey do we have any good reason to keep byte mode support at all? Thinking that we may just have the controller set in FIFO mode regardless of CONFIG_UART_INTERRUPT_DRIVEN, so that we don't have to have two different poll implementation to maintain and test for the poll API, as well as a change of behavior if the interrupt mode gets compiled out.

NPCX5 and NPCX7 series support byte mode. I don't think it's our place to decide what consumers of Zephyr will need.

Commit:
uart-npcx: re-organize interrupt driven functions.

Simple reorg of the function declarations. This change moves
the interrupt driven functions declared at the bottom of the
uart_npcx.c file to the top, clustered under the same ifdef.

Signed-off-by: Yuval Peress <peress@chromium.org>

dcpleung commented on drivers/serial/uart_npcx.c:
```diff
@@ -257,7 +213,100 @@ static void uart_npcx_isr(const struct device *dev)
+}
+#endif /* !CONFIG_UART_INTERRUPT_DRIVEN */
+
+/* UART api functions */
+static int uart_npcx_poll_in(const struct device *dev, unsigned char *c)
+{
+	struct uart_reg *const inst = HAL_INSTANCE(dev);
+
+	/* Rx single byte buffer is not full */
+	if (!IS_BIT_SET(inst->UICTRL, NPCX_UICTRL_RBF))
+		return -1;
+
+	*c = inst->URBUF;
+	return 0;
+}
+
+static void uart_npcx_poll_out(const struct device *dev, unsigned char c)
+{
+	struct uart_reg *const inst = HAL_INSTANCE(dev);
+
+	/* Wait while Tx single byte buffer is ready to send */
+	while (!IS_BIT_SET(inst->UICTRL, NPCX_UICTRL_TBE))
+		continue;
+
+	inst->UTBUF = c;
+}
```
Please remove the duplicate functions here, or else you would have two poll_in and two poll_out functions.

yperess replied:
@galak Requested that I duplicate the whole function. Frustrating that GitHub doesn't keep the history, but the original version was:
```
static void uart_npcx_poll_in(...)
{
#ifdef CONFIG_UART_INTERRUPT_DRIVEN
    // Call fifo function
#else
  // Do the same as before
#endif
}
```

dcpleung replied:
The one marked here are outside of the #ifdef. So regardless of `CONFIG_UART_INTERRUPT_DRIVEN`, there are two copies of poll_in() and poll_out(). The ones after `#endif` need to be removed.

yperess replied:
Oh, I didn't even notice that when I did the rebase after splitting out the change. Fixed.
